### PR TITLE
[JENKINS-33843] Permit disabling optional dependencies

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-33843](https://issues.jenkins-ci.org/browse/JENKINS-33843).

- [X] fixed data model bug involving implicit optional dependencies on detached plugins
- [X] introduced and used alternate methods considering only mandatory dependencies
- [x] added a bit of test coverage

It would be possible to improve the GUI to note unsatisfied optional dependencies somehow, given data such as this:

```groovy
Jenkins.instance.pluginManager.plugins.sort(false, {it.shortName}).each {p ->
  p.optionalDependencies*.shortName.each {d ->
    def dp = Jenkins.instance.pluginManager.getPlugin(d);
    if (dp == null) {
      println("missing dep $p.shortName → $d");
    } else if (!dp.enabled) {
      println("disabled dep $p.shortName → $d");
    }
  };
}; null
```

Ideally this would also be reactive, meaning `_table.js` would update it as you click **Enabled** checkboxes. Perhaps missing deps would be displayed under the depender, whereas disabled deps would be displayed under the dependee. I decided not to spend time touching the GUI in this PR, mainly because I am not very good at it; would be a good enhancement for someone else to pick up.

### Proposed changelog entries

* The plugin manager’s **Installed** tab incorrectly prevented administrators from disabling a plugin if any other plugin had even an optional dependency on it.

### Desired reviewers

@jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: Unfortunately there was duplicated information here (a field and also membership in a collection) which had drifted out of synch.

File: core/src/main/resources/hudson/PluginManager/installed.jelly
Comment (jglick) [inline core/src/main/resources/hudson/PluginManager/installed.jelly]: See PR description: GUI no longer deals with optional dependencies at all.

File: test/src/test/java/hudson/PluginManagerInstalledGUITest.java
Comment (jglick) [inline test/src/test/java/hudson/PluginManagerInstalledGUITest.java]: Original test case assumed that the implicit `tasks` → detached `cvs` dep could be used to test everything else. Was not rich enough data.

**Comment by daniel-beck** [comment]
> implicit optional dependencies on detached plugins

If we introduce a distinction between optional and mandatory dependencies on the UI, I would prefer we treat implied dependencies as mandatory rather than optional for that, to prevent inexperienced users from shooting themselves in the foot.

While recent detachments were pretty boring, we also had something like `junit`.

**Review by oleg-nenashev — APPROVED** [review]
LGTM after a quick review. I have not tested it locally, but we have enough time to fix/revert logic before it gets into LTS. Also, last weekly already had low community ratings, so there is a low expectation

**Comment by jglick** [comment]
> I would prefer we treat implied dependencies as mandatory

I discussed this topic [in JIRA](https://issues.jenkins-ci.org/browse/JENKINS-33843?focusedCommentId=365600&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-365600) and on balance think it is preferable to keep them as `optional`.

**Comment by daniel-beck** [comment]
Could implied dependencies be treated as a third kind of dependency wrt. the UI relatively easily? We want to treat them as optional wrt. plugin loading, but on the UI specifically, since we don't know whether they're needed or not, we should add appropriate warnings for users. There's a difference between an optional dependency that can be disabled (unless there's a bug in the depending plugin), and an implied dependency where we do not know.

We might also be able to enhance the exception error message when an `@Extension` fails to load to print the optional and implied dependencies of the affected plugin that are currently disabled. This could help users diagnose the problem in the above situations (plugin bug and needed implied dependency).

**Review by varyvol — APPROVED** [review]

**Review by fcojfernandez — APPROVED** [review]

**Comment by jglick** [comment]
> Could implied dependencies be treated as a third kind of dependency wrt. the UI relatively easily?

I am leery of making the dependency system and associated UI even more complex than it already is. (Really I would like to kill optional dependencies, but that would require some broader UX changes.)

> an optional dependency that can be disabled (unless there's a bug in the depending plugin)

In practice there usually _is_ a bug in the depending plugin. More often than that a plugin is actually using a detached dependency, I think.

If you think the option in the PR description

> It would be possible to improve the GUI to note unsatisfied optional dependencies somehow

is a prerequisite—i.e., that we would be _worse off_ letting users disable detached plugins freely than the current situation in which they are unable to disable many plugins from the GUI and we sometimes guide them to create `*.jpi.disabled` files manually as a workaround—then please say so. Clearly there is a tradeoff here. My inclination is to let admins do what they need to do, and in what I suspect are the relatively uncommon cases that a detached dependency was actually mandatory, and we find out about it, the fix is straightforward: just update `jenkins.version` (forcing the maintainer to add the dep), and release.

> We might also be able to enhance the exception error message when…

Yes, this would be useful.

**Comment by oleg-nenashev** [comment]
@daniel-beck are you fine with the provided feedback?

**Comment by daniel-beck** [comment]
> I am leery of making the dependency system and associated UI even more complex than it already is. (Really I would like to kill optional dependencies, but that would require some broader UX changes.)

It would simply reflect on the UI that for implied dependencies, we do not know whether they're mandatory (possibly), optional (unlikely), or not used at all.

> If you think the option in the PR description
> 
> > It would be possible to improve the GUI to note unsatisfied optional dependencies somehow
> 
> is a prerequisite—i.e., that we would be _worse off_ letting users disable detached plugins freely than the current situation in which they are unable to disable many plugins from the GUI and we sometimes guide them to create `*.jpi.disabled` files manually as a workaround—then please say so. Clearly there is a tradeoff here.

I expect this to be the case. Notably, the resulting potential problem is much worse if it occurs.

**Comment by jglick** [comment]
> reflect on the UI that for implied dependencies, we do not know whether…

One relatively simple change that would not require messing around with internal APIs would be to have the UI track plugins which are in the detached list (whether or not you have installed the originally detached version), and include a warning in the hover tip that appears over the enablement checkbox that disabling the plugin might cause errors in some of the listed dependents (without calling out specific ones). We would only need to add something like `PluginWrapper.isDetached()` which checks `DetachedPluginsUtil.isDetachedPlugin`. Would that alleviate your fear?

**Comment by daniel-beck** [comment]
> Would that alleviate your fear?

I think that would be a good compromise 👍

**Pull request merged**
